File: ocmsdk/src/main/java/com/gigigo/orchextra/wrapper/OxManagerImpl.java
package com.gigigo.orchextra.wrapper;

import android.app.Application;
import com.gigigo.imagerecognitioninterface.ImageRecognitionCredentials;
import com.gigigo.orchextra.CustomSchemeReceiver;
import com.gigigo.orchextra.Orchextra;
import com.gigigo.orchextra.OrchextraBuilder;
import com.gigigo.orchextra.OrchextraLogLevel;
import com.gigigo.orchextra.device.bluetooth.beacons.BeaconBackgroundModeScan;
import com.gigigo.orchextra.ocm.callbacks.OnCustomSchemeReceiver;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import orchextra.javax.inject.Inject;

/**
 * Created by alex on 01/12/2017.
 */

public class OxManagerImpl implements OxManager {

  private OnCustomSchemeReceiver onCustomSchemeReceiver;
  private HashMap<CrmUser.Gender, com.gigigo.orchextra.CrmUser.Gender> genders;

  //cambio para el inicio selectivo, MEJORAR,
  //necesitamos un contexto para q la funcion setNewOrchextracredentials pueda comprobar las preferences
  //lo suyo es no guardarlo en las preferences, de momneto así y una mejora sencilla seria añadir el contexto a
  //la funcion de setNewOrchextracredentials, para no mantener el application cuando no es necesario
  private CustomSchemeReceiver onOxCustomSchemeReceiver =
      customScheme -> callOnCustomSchemeReceiver(customScheme);

  @Inject
  public OxManagerImpl() {
    genders = new HashMap<>();
    genders.put(CrmUser.Gender.GenderFemale, com.gigigo.orchextra.CrmUser.Gender.GenderFemale);
    genders.put(CrmUser.Gender.GenderMale, com.gigigo.orchextra.CrmUser.Gender.GenderMale);
    genders.put(CrmUser.Gender.GenderND, com.gigigo.orchextra.CrmUser.Gender.GenderND);
  }

  @Override public void callOnCustomSchemeReceiver(String customScheme) {
    if (onCustomSchemeReceiver != null) {
      onCustomSchemeReceiver.onReceive(customScheme);
    }
  }

  @Override public void startImageRecognition() {
    Orchextra.startImageRecognition();
  }

  @Override public void startScanner() {
    Orchextra.startScannerActivity();
  }

  @Override public void init(Application app, Config config) {

    OrchextraBuilder builder = new OrchextraBuilder(app);
    builder.setApiKeyAndSecret(config.getApiKey(), config.getApiSecret())
        .setLogLevel(OrchextraLogLevel.NETWORK)
        .setBackgroundBeaconScanMode(BeaconBackgroundModeScan.NORMAL)
        .setOrchextraCompletionCallback(new com.gigigo.orchextra.OrchextraCompletionCallback() {
          @Override public void onSuccess() {
            config.getOrchextraCompletionCallback().onSuccess();
          }

          @Override public void onError(String message) {
            config.getOrchextraCompletionCallback().onError(message);
          }

          @Override public void onInit(String s) {
            config.getOrchextraCompletionCallback().onInit(s);
          }

          @Override public void onConfigurationReceive(String s) {
            config.getOrchextraCompletionCallback().onConfigurationReceive(s);
          }
        });

    if (config.getNotificationActivityClass() != null) {
      builder.setNotificationActivityClass(config.getNotificationActivityClass().toString());
    }
    if (config.getSenderId() != null && !config.getSenderId().equals("")) {
      builder.setGcmSenderId(config.getSenderId());
    }
    if (config.getVuforia() != null) {
      builder.setImageRecognitionModule(
          new com.gigigo.imagerecognitioninterface.ImageRecognition() {
            @Override public <T> void setContextProvider(T contextProvider) {
              config.getVuforia().setContextProvider(contextProvider);
            }

            @Override
            public void startImageRecognition(
                ImageRecognitionCredentials imageRecognitionCredentials) {
              config.getVuforia().startImageRecognition(
                  new com.gigigo.orchextra.wrapper.ImageRecognitionCredentials() {
                    @Override public String getClientAccessKey() {
                      return imageRecognitionCredentials.getClientAccessKey();
                    }

                    @Override public String getLicensekey() {
                      return imageRecognitionCredentials.getLicensekey();
                    }

                    @Override public String getClientSecretKey() {
                      return imageRecognitionCredentials.getClientSecretKey();
                    }
                  });
            }
          });
    }

    Orchextra.initialize(builder);

    Orchextra.setCustomSchemeReceiver(onOxCustomSchemeReceiver);
  }

  @Override public void getToken() {

  }

  @Override public void bindUser(CrmUser crmUser) {
    com.gigigo.orchextra.CrmUser crmUserOx = new com.gigigo.orchextra.CrmUser(crmUser.getCrmId(),
        crmUser.getBirthdate(), genders.get(crmUser.getGender()));

    Orchextra.bindUser(crmUserOx);
    Orchextra.commitConfiguration();
  }

  @Override public void unBindUser() {

  }

  @Override public void bindDevice(String businessUnit) {
    List<String> bussinessUnits = new ArrayList();
    bussinessUnits.add(businessUnit);
    Orchextra.setDeviceBusinessUnits(bussinessUnits);
  }

  @Override public void unBindDevice() {

  }

  @Override public void setOnCustomSchemeReceiver(OnCustomSchemeReceiver onCustomSchemeReceiver) {
    this.onCustomSchemeReceiver = onCustomSchemeReceiver;
  }

  @Override public void start() {
    Orchextra.start();
  }

  @Override public void stop() {
    Orchextra.stop();
  }

  @Override
  public void updateSDKCredentials(String apiKey, String apiSecret, boolean forceCallback) {
    Orchextra.updateSDKCredentials(apiKey, apiSecret, forceCallback);
  }
}

File: ocmsdk/src/main/java/com/gigigo/orchextra/wrapper/OxManagerImpl.kt
package com.gigigo.orchextra.wrapper

import android.app.Application
import com.gigigo.orchextra.core.Orchextra
import com.gigigo.orchextra.core.OrchextraErrorListener
import com.gigigo.orchextra.core.OrchextraOptions
import com.gigigo.orchextra.core.OrchextraStatusListener
import com.gigigo.orchextra.core.domain.entities.Error
import com.gigigo.orchextra.core.domain.entities.OxCRM
import com.gigigo.orchextra.geofence.OxGeofenceImp
import com.gigigo.orchextra.indoorpositioning.OxIndoorPositioningImp
import com.gigigo.orchextra.ocm.callbacks.OnCustomSchemeReceiver
import com.gigigo.orchextra.scanner.OxScannerImp
import com.gigigo.orchextra.wrapper.OxManager.ErrorListener
import com.gigigo.orchextra.wrapper.OxManager.ScanCodeListener
import com.gigigo.orchextra.wrapper.OxManager.StatusListener
import com.gigigo.orchextra.wrapper.OxManager.TokenReceiver

class OxManagerImpl : OxManager {

  private val orchextra = Orchextra
  private var customSchemeReceiver: OnCustomSchemeReceiver? = null
  private val genders: HashMap<CrmUser.Gender, String> = HashMap()

  init {
    genders[CrmUser.Gender.GenderFemale] = "female"
    genders[CrmUser.Gender.GenderMale] = "male"
    genders[CrmUser.Gender.GenderND] = "nd"
  }

  override fun startImageRecognition() = orchextra.openImageRecognition()

  override fun startScanner() = orchextra.openScanner()

  override fun scanCode(scanCodeListener: ScanCodeListener) = with(orchextra) {
    scanCode { scanCodeListener.onCodeScan(it) }
  }

  override fun init(application: Application, config: OxConfig, statusListener: StatusListener) {

    orchextra.setStatusListener(object : OrchextraStatusListener {
      override fun onStatusChange(isReady: Boolean) {
        if (isReady) {
          orchextra.getTriggerManager().scanner = OxScannerImp.create(application)
          orchextra.getTriggerManager().geofence = OxGeofenceImp.create(application)
          orchextra.getTriggerManager().indoorPositioning = OxIndoorPositioningImp.create(
              application)

          config.notificationActivityClass?.let {
            orchextra.setNotificationActivityClass(it)
          }

          statusListener.onSuccess()
        } else {
          statusListener.onError("SDK isn't ready")
        }
      }
    })

    orchextra.setErrorListener(object : OrchextraErrorListener {
      override fun onError(error: Error) {
        statusListener.onError(error.message)
      }
    })

    val options = OrchextraOptions.Builder().firebaseApiKey(config.firebaseApiKey)
        .firebaseApplicationId(config.firebaseApplicationId)
        .triggeringEnabled(config.triggeringEnabled)
        .anonymous(config.anonymous)
        .deviceBusinessUnits(config.deviceBusinessUnits)
        .debuggable(true)

        .build()

    orchextra.init(application, config.apiKey, config.apiSecret, options)
    orchextra.setScanTime(30)
  }

  override fun finish() = orchextra.finish()

  override fun removeListeners() = with(orchextra) {
    removeStatusListener()
    removeErrorListener()
  }

  override fun isReady(): Boolean = orchextra.isReady()

  override fun getToken(tokenReceiver: TokenReceiver) =
      orchextra.getToken { token -> tokenReceiver.onGetToken(token) }

  override fun setErrorListener(errorListener: ErrorListener) {
    orchextra.setErrorListener(object : OrchextraErrorListener {
      override fun onError(error: Error) {
        errorListener.onError(error.message)
      }
    })
  }

  override fun setBusinessUnits(businessUnits: List<String>, statusListener: StatusListener) {
    orchextra.getCrmManager().setDeviceData(null, businessUnits) { statusListener.onSuccess() }
  }

  override fun bindUser(crmUser: CrmUser, statusListener: StatusListener) {
    val crm = OxCRM(crmUser.crmId, crmUser.gender?.name, crmUser.birthdate?.time)
    orchextra.getCrmManager().bindUser(crm) { statusListener.onSuccess() }
  }

  override fun unBindUser(statusListener: StatusListener) {
    orchextra.getCrmManager().unbindUser { statusListener.onSuccess() }
  }

  override fun setCustomSchemeReceiver(customSchemeReceiver: OnCustomSchemeReceiver) {
    this.customSchemeReceiver = customSchemeReceiver
    orchextra.setCustomActionListener { customSchemeReceiver.onReceive(it) }
  }

  override fun onCustomScheme(customScheme: String) {
    customSchemeReceiver?.onReceive(customScheme)
  }
}